### ISSUE TITLE
[Issue]: Intel ARC broken with 2.0 Pytorch

### ISSUE THREAD

**Issue by waterlubber** [vladmandic/automatic#1942]
### Issue Description

I'm using an Arc A770 on Arch Linux. The latest commits to the repository appear to have broken the installation, which was working about a month ago when I set it up with Pytorch 1.x

After troubleshooting installation last night (there appears to have been an issue where pip did not recognize the proper wheels in Intel's repository, and I had to manually install them) I wiped the repo and made a fresh clone, then installed with 

`./webui.sh --use-ipex`.

Installation succeeded, but I could not launch due to the following error:
`ImportError: [snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/lib/libintel-ext-pt-gpu.so: undefined symbol: _ZN4sycl3_V16detail14tls_code_loc_tD1Ev`

It appears to be some sort of ABI conflict. When experimenting with the Intel wheels, they appear to be compiled for a "cxx 11 ABI": I am not sure exactly what *other* programs need to be compiled with this ABI, but it's probably a distribution conflict.

What can I do to resolve this? I'd hate to use a docker container and definitely won't be dual booting another distro just for this. Should I recompile Intel's pytorch libraries locally? I'm not particularly familiar with Python's environment (every time I've interacted with it, it's been miserable - I don't use Python very often because of that).

As an aside: on previous versions, I had commented out ipexrun in ./webui.sh; it caused strange core dumps and other errors. 

There are no issues with the GPU-side configuration to my knowledge; until my recent update, things were working fairly well (except for training, which did not work, and numerical stability issues at large batch sizes or resolutions). 

Please let me know how I can help, or if I should move this issue to the Intel people. 

### Version Platform Description

Python 3.10
Arch Linux
`Linux gamingPC 6.4.8-zen1-1-zen #1 ZEN SMP PREEMPT_DYNAMIC Thu, 03 Aug 2023 16:01:52 +0000 x86_64 GNU/Linux`
Intel Arc A770

### Relevant log output

```shell
Create and activate python venv
Setting OneAPI environment
 
:: initializing oneAPI environment ...
   webui.sh: BASH_VERSION = 5.1.16(1)-release
   args: Using "$@" for setvars.sh arguments: --use-ipex
:: compiler -- latest
:: mkl -- latest
:: tbb -- latest
:: oneAPI environment initialized ::
 
Launching ipexrun launch.py...
[snip]/automatic/venv/lib/python3.10/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: ''If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
Traceback (most recent call last):
  File "[snip]/automatic/venv/bin/ipexrun", line 5, in <module>
    from intel_extension_for_pytorch.launcher import main
  File "[snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/__init__.py", line 93, in <module>
    from .utils._proxy_module import *
  File "[snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/utils/_proxy_module.py", line 2, in <module>
    import intel_extension_for_pytorch._C
ImportError: [snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/lib/libintel-ext-pt-gpu.so: undefined symbol: _ZN4sycl3_V16detail14tls_code_loc_tD1Ev
```

### Acknowledgements

- [X] I have read the above and searched for existing issues
- [X] I confirm that this is classified correctly and its not an extension or diffusers-specific issue

**Comment by vladmandic** [comment]
cc @Disty0

**Comment by Disty0** [comment]
Update your OneAPI Basekit drivers.

**Comment by waterlubber** [comment]
Alright, I installed Arch's `intel-oneapi-basekit` package from `extra-staging` (I am wary of just running an out-of-tree installer as-is), and got a wholly different error:

```

Create and activate python venv
Setting OneAPI environment
 
:: initializing oneAPI environment ...
   webui.sh: BASH_VERSION = 5.1.16(1)-release
   args: Using "$@" for setvars.sh arguments: --use-ipex
:: advisor -- latest
:: ccl -- latest
:: compiler -- latest
:: dal -- latest
:: debugger -- latest
:: dev-utilities -- latest
:: dnnl -- latest
:: dpcpp-ct -- latest
:: dpl -- latest
:: ipp -- latest
:: ippcp -- latest
:: ipp -- latest
:: mkl -- latest
:: mpi -- latest
:: tbb -- latest
:: vtune -- latest
:: oneAPI environment initialized ::
 
Launching ipexrun launch.py...
[snip]/automatic/venv/lib/python3.10/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: ''If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
2023-08-05 09:47:36.230211: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-08-05 09:47:36.679697: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-08-05 09:47:36.679963: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-08-05 09:47:37.691239: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2023-08-05 09:47:41.328854: I itex/core/wrapper/itex_gpu_wrapper.cc:35] Intel Extension for Tensorflow* GPU backend is loaded.
2023-08-05 09:47:41.391779: W itex/core/ops/op_init.cc:58] Op: _QuantizedMaxPool3D is already registered in Tensorflow
2023-08-05 09:47:41.410956: I itex/core/devices/gpu/itex_gpu_runtime.cc:129] Selected platform: Intel(R) Level-Zero
2023-08-05 09:47:41.411265: I itex/core/devices/gpu/itex_gpu_runtime.cc:154] number of sub-devices is zero, expose root device.
2023-08-05 09:47:41,706 - numexpr.utils - INFO - Note: NumExpr detected 12 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
2023-08-05 09:47:41,706 - numexpr.utils - INFO - NumExpr defaulting to 8 threads.
[snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/launcher.py:102: UserWarning: Backend is not specified, it will automatically default to cpu.
  warnings.warn(
2023-08-05 09:47:42,155 - intel_extension_for_pytorch.cpu.launch.launch - INFO - Use 'auto' => 'tcmalloc' memory allocator.
2023-08-05 09:47:42,155 - intel_extension_for_pytorch.cpu.launch.launch - WARNING - 'intel' OpenMP runtime is not found in ['[snip]/automatic/venv/lib/', '/home/waterlubber/.local/lib/', '/usr/local/lib/', '/usr/local/lib64/', '/usr/lib/', '/usr/lib64/', '/usr/lib/x86_64-linux-gnu/'].
2023-08-05 09:47:42,156 - intel_extension_for_pytorch.cpu.launch.launch - INFO - Use 'default' OpenMP runtime.
2023-08-05 09:47:42,157 - intel_extension_for_pytorch.cpu.launch.launch - INFO - Use 'auto' => 'numactl' multi-task manager.
2023-08-05 09:47:42,157 - intel_extension_for_pytorch.cpu.launch.launch - INFO - env: Untouched preset environment variables are not displayed.
2023-08-05 09:47:42,158 - intel_extension_for_pytorch.cpu.launch.launch - INFO - env: LD_PRELOAD=/usr/lib/libtcmalloc.so
2023-08-05 09:47:42,158 - intel_extension_for_pytorch.cpu.launch.launch - INFO - env: OMP_SCHEDULE=STATIC
2023-08-05 09:47:42,158 - intel_extension_for_pytorch.cpu.launch.launch - INFO - env: OMP_PROC_BIND=CLOSE
2023-08-05 09:47:42,158 - intel_extension_for_pytorch.cpu.launch.launch - INFO - env: OMP_NUM_THREADS=6
2023-08-05 09:47:42,158 - intel_extension_for_pytorch.cpu.launch.launch - INFO - cmd: numactl -C 0-5 -m 0 [snip]/automatic/venv/bin/python3.10 -u launch.py --use-ipex
09:47:42-367553 INFO     Starting SD.Next                                                                                                                             
09:47:42-371302 INFO     Python 3.10.12 on Linux                                                                                                                      
09:47:42-405568 INFO     Version: 4b01e607 Fri Aug 4 19:05:25 2023 +0200                                                                                              
09:47:42-506731 INFO     Latest published version: cfe14884a038d0e3e26c9432055bf23715c2d54a 2023-08-05T12:06:56Z                                                      
09:47:42-508182 INFO     Intel OneAPI Toolkit detected                                                                                                                
09:47:42-559835 WARNING  Modified files: ['webui.sh']                                                                                                                 
09:47:42-570517 INFO     Disabled extensions: []                                                                                                                      
09:47:42-571597 INFO     Enabled extensions-builtin: ['LDSR', 'Lora', 'ScuNET', 'SwinIR', 'a1111-sd-webui-lycoris', 'clip-interrogator-ext',                          
                         'multidiffusion-upscaler-for-automatic1111', 'sd-dynamic-thresholding', 'sd-extension-system-info', 'sd-webui-agent-scheduler',              
                         'sd-webui-controlnet', 'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg']                                              
09:47:42-575184 INFO     Enabled extensions: []                                                                                                                       
09:47:42-576129 INFO     No changes detected: Quick launch active                                                                                                     
09:47:42-576954 INFO     Verifying requirements                                                                                                                       
09:47:42-592297 INFO     Disabled extensions: []                                                                                                                      
09:47:42-593402 INFO     Enabled extensions-builtin: ['LDSR', 'Lora', 'ScuNET', 'SwinIR', 'a1111-sd-webui-lycoris', 'clip-interrogator-ext',                          
                         'multidiffusion-upscaler-for-automatic1111', 'sd-dynamic-thresholding', 'sd-extension-system-info', 'sd-webui-agent-scheduler',              
                         'sd-webui-controlnet', 'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg']                                              
09:47:42-595317 INFO     Enabled extensions: []                                                                                                                       
09:47:42-598953 INFO     Extension preload: 0.0s [snip]/automatic/extensions-builtin                                                                 
09:47:42-599915 INFO     Extension preload: 0.0s [snip]/automatic/extensions                                                                         
09:47:42-608233 INFO     Server arguments: ['--use-ipex']                                                                                                             
[snip]/automatic/venv/lib/python3.10/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: ''If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
free(): invalid pointer
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ [snip]/automatic/venv/bin/ipexrun:8 in <module>                                 │
│                                                                                                  │
│   5 from intel_extension_for_pytorch.launcher import main                                        │
│   6 if __name__ == '__main__':                                                                   │
│   7 │   sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])                         │
│ ❱ 8 │   sys.exit(main())                                                                         │
│   9                                                                                              │
│                                                                                                  │
│ [snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/ │
│ launcher.py:109 in main                                                                          │
│                                                                                                  │
│   106 │   │   sys.argv.insert(1, "cpu")                                                          │
│   107 │   args = parser.parse_args()                                                             │
│   108 │   if args.backend == "cpu":                                                              │
│ ❱ 109 │   │   cpu_run_main_with_args(args)                                                       │
│   110 │   elif args.backend == "xpu":                                                            │
│   111 │   │   xpu_run_main_with_args(args)                                                       │
│   112 │   else:                                                                                  │
│                                                                                                  │
│ [snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/ │
│ cpu/launch/launch.py:453 in run_main_with_args                                                   │
│                                                                                                  │
│   450 │   else:                                                                                  │
│   451 │   │   launcher = launcher_multi_instances                                                │
│   452 │                                                                                          │
│ ❱ 453 │   launcher.launch(args)                                                                  │
│   454 │   for x in sorted(set(os.environ.keys()) - env_before):                                  │
│   455 │   │   logger.debug(f"{x}={os.environ[x]}")                                               │
│   456                                                                                            │
│                                                                                                  │
│ [snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/ │
│ cpu/launch/launcher_multi_instances.py:316 in launch                                             │
│                                                                                                  │
│   313 │   │   │   │   p = process["process"]                                                     │
│   314 │   │   │   │   p.wait()                                                                   │
│   315 │   │   │   │   if p.returncode != 0:                                                      │
│ ❱ 316 │   │   │   │   │   raise subprocess.CalledProcessError(                                   │
│   317 │   │   │   │   │   │   returncode=p.returncode, cmd=process["cmd"]                        │
│   318 │   │   │   │   │   )                                                                      │
│   319 │   │   finally:                                                                           │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
CalledProcessError: Command 'numactl -C 0-5 -m 0 [snip]/automatic/venv/bin/python3.10 -u launch.py --use-ipex' died with <Signals.SIGABRT: 6>.
```
If I comment out the `ipexrun` in `webui.sh` (as I had to do before) it starts normally and everything works again.
If you would like, I can create a second issue for ipexrun breaking things on Arch systems; perhaps it would be better to wait for Intel's libraries to percolate downstream first (as I'm currently running a wretched partially upgraded system)

**Comment by Disty0** [comment]
I also use Arch Linux and it runs fine on my system...
But webui.sh now checks for `--use-ipex` to use ipexrun:
https://github.com/vladmandic/automatic/commit/489d0382cfd95ba4420d94dce4d2432c0cb54dd8

You can start the WebUI without `--use-ipex` and ipexrun won't be in use.
IPEX is auto detect so WebUI will detect IPEX without `--use-ipex` too.

**Comment by waterlubber** [comment]
What nonstandard packages are you running? You mentioned installing with Intel's installer. 

I've also been experiencing issues with both training and inference; both of them give me a variety of errors about F64 values appearing, or `RuntimeError: expected scalar type Float but found Half stable diffusion`. Going to try resetting configs and trying again.

**Comment by Disty0** [comment]
IPEX 2.0 doesn't have the GPU version of GradScaler so training won't work without patching.
Second issue seems to be this:
https://github.com/vladmandic/automatic/issues/1953

Packages i've installed:
```
yay -S python310 git unzip python-pip python-virtualenv numactl intel-media-driver intel-oneapi-basekit intel-compute-runtime intel-graphics-compiler 
```

**Comment by waterlubber** [comment]
You mentioned having a more updated basekit than what's in the repositories - are the packages you mentioned the versions in extra, or extra-staging? I can update to the staging versions if necessary.

**Comment by Disty0** [comment]
OneAPI Basekit in extra-staging repo didn't exist when i wrote the first message.
I installed OneAPI Basekit 2023.2 from extra-staging, removed the standalone one and it's working fine on my end.

**Comment by waterlubber** [comment]
Thank you very much for the help and patches. BF16 is working again and some of the performance regressions I've experienced since 1.13 have been reduced.

Do you know if patches for GPU Gradscaler on IPEX2.0 are out yet, or when Intel might implement it? 

I have all of those packages installed; all are the versions in the Arch `extra` repo with the aforementioned exception of `intel-oneapi-basekit`, which is version 2023.2.0.49397-1. I installed this package by directly downloading from the mirror, with `-U`. Did you change your system repository config, perhaps? It seems weird that two otherwise identical systems would produce different results like this. Do you know if there are any configuration files that might be floating around to cause this difference, perhaps ones created by the Intel installer on your machine?



Another issue I've been having is related to the compiler settings here:
![image](https://github.com/vladmandic/automatic/assets/7517025/839b6207-2b1d-479e-92a9-aeb8503e4acb)
This is an experimental feature and it's not a big deal if it isn't working, but it has regressed on my machine. Perhaps its just some missing feature. 
```
13:59:55-559697 ERROR    WON'T CONVERT get_available_vram [snip]/automatic/modules/sd_hijack_optimizations.py line 31                                
     due to:                                                                                                                                      
     Traceback (most recent call last):                                                                                                           
       File "[snip]/automatic/venv/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 71, in unimplemented                 
         raise Unsupported(msg)                                                                                                                   
     torch._dynamo.exc.Unsupported: call_function memory_stats in skip_files                                                                      
     [snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/xpu/memory.py                                
                                                                                                                                                  
     from user code:                                                                                                                              
        File "[snip]/automatic/modules/sd_hijack_optimizations.py", line 34, in get_available_vram                               
         stats = torch.cuda.memory_stats(shared.device)                                                                                           
                                                                                                                                                  
     Set torch._dynamo.config.verbose=True for more information                                                                                   
                                                                                                                                                  
                                                                                                                                                  
13:59:55-559697 ERROR    WON'T CONVERT get_available_vram [snip]/automatic/modules/sd_hijack_optimizations.py line 31                                
     due to:                                                                                                                                      
     Traceback (most recent call last):                                                                                                           
       File "[snip]/automatic/venv/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 71, in unimplemented                 
         raise Unsupported(msg)                                                                                                                   
     torch._dynamo.exc.Unsupported: call_function memory_stats in skip_files                                                                      
     [snip]/automatic/venv/lib/python3.10/site-packages/intel_extension_for_pytorch/xpu/memory.py                                
                                                                                                                                                  
     from user code:                                                                                                                              
        File "[snip]/automatic/modules/sd_hijack_optimizations.py", line 34, in get_available_vram                               
         stats = torch.cuda.memory_stats(shared.device)                                                                                           
                                                                                                                                                  
     Set torch._dynamo.config.verbose=True for more information                                                                                   
                                                                                                                                                  
                                                                                                                                                  
13:59:55-568894 ERROR    WON'T CONVERT <graph break in <lambda>> [snip]/automatic/modules/ipex_specific/__init__.py line 82                          
     due to:                                                                                                                                      
     Traceback (most recent call last):                                                                                                           
       File "[snip]/automatic/venv/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 114, in __init__     
         assert isinstance(                                                                                                                       
     AssertionError: expected FunctionType found instancemethod <instancemethod  at 0x7f9040d4b130>                                               
                                                                                                                                                  
     from user code:                                                                                                                              
        File "[snip]/automatic/modules/ipex_specific/__init__.py", line 82, in <graph break in <lambda>>                         
         torch.cuda.mem_get_info = lambda device=None: [(torch.xpu.get_device_properties(device).total_memory -                                   
     torch.xpu.memory_allocated(device)), torch.xpu.get_device_properties(device).total_memory]                                                   
                                                                                                                                                  
     Set torch._dynamo.config.verbose=True for more information                                                                                   
                                                                                                                                                  
                                                                                                                                                  
13:59:55-568894 ERROR    WON'T CONVERT <graph break in <lambda>> [snip]/automatic/modules/ipex_specific/__init__.py line 82                          
     due to:                                                                                                                                      
     Traceback (most recent call last):                                                                                                           
       File "[snip]/automatic/venv/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py", line 114, in __init__     
         assert isinstance(                                                                                                                       
     AssertionError: expected FunctionType found instancemethod <instancemethod  at 0x7f9040d4b130>                                               
                                                                                                                                                  
     from user code:                                                                                                                              
        File "[snip]/automatic/modules/ipex_specific/__init__.py", line 82, in <graph break in <lambda>>                         
         torch.cuda.mem_get_info = lambda device=None: [(torch.xpu.get_device_properties(device).total_memory -                                   
     torch.xpu.memory_allocated(device)), torch.xpu.get_device_properties(device).total_memory]                                                   
                                                                                                                                                  
     Set torch._dynamo.config.verbose=True for more information       
```

Finally (thank you very much for your work on this project and help with this): What does 'IPEX Optimize for Intel GPUs' do? Doesn't seem to make much of a difference on my machine.

**Comment by Disty0** [comment]
> Do you know if patches for GPU Gradscaler on IPEX2.0 are out yet, or when Intel might implement it?~ Found [the patch. ](https://github.com/Disty0/ipex_grad_scaler), although I wasn't able to get it working ("could not create a primitive descriptor for a matmul primitive")

I patched 1.13 manually before, i don't know anyone else who patches it.
It's still not in the xpu-master branch so official support won't be in at least a month.
I don't know if they have any plans at all.


> I have all of those packages installed; all are the versions in the Arch `extra` repo with the aforementioned exception of `intel-oneapi-basekit`, which is version 2023.2.0.49397-1. I installed this package by directly downloading from the mirror, with `-U`. Did you change your system repository config, perhaps? It seems weird that two otherwise identical systems would produce different results like this. Do you know if there are any configuration files that might be floating around to cause this difference, perhaps ones created by the Intel installer on your machine?

Completely removed the standalone basekit, then installed it like how you did and it works fine.
Also output of `yay -s Intel | grep "(Installed)"`
```
58 multilib/lib32-libva-intel-driver 2.4.1-1 (1.1 MiB 7.9 MiB) (Installed)
57 multilib/lib32-vulkan-intel 1:23.1.5-1 (3.7 MiB 27.9 MiB) (Installed)
54 extra/libva-intel-driver 2.4.1-2 (1.0 MiB 7.8 MiB) (Installed)
47 extra/libva-utils 2.19.0-1 (524.8 KiB 3.1 MiB) (Installed)
42 extra/openimagedenoise 1.4.3-1 (43.1 MiB 49.1 MiB) (Installed)
41 extra/libmfx 23.2.2-2 (57.4 KiB 285.9 KiB) (Installed)
39 extra/openpgl 0.5.0-5 (221.4 KiB 854.3 KiB) (Installed)
24 extra/nvtop 3.0.2-1 (62.1 KiB 157.0 KiB) (Installed)
22 extra/igsc 0.8.9-1 (64.9 KiB 222.4 KiB) (Installed)
21 extra/intel-compute-runtime 23.22.26516.18-1 (12.0 MiB 72.5 MiB) (Installed)
19 extra/intel-graphics-compiler 1:1.0.14062.11-1 (9.8 MiB 45.3 MiB) (Installed)
18 extra/intel-opencl-clang 15.0.0-1 (98.5 KiB 1009.4 KiB) (Installed)
12 extra/intel-media-driver 23.3.0-1 (6.6 MiB 36.0 MiB) (Installed)
11 extra/intel-gpu-tools 1.27-2 (4.1 MiB 25.5 MiB) (Installed)
6 extra/onednn 3.1.1-1 (7.1 MiB 36.9 MiB) (Installed)
5 extra/onetbb 2021.9.0-1 (371.6 KiB 1.9 MiB) (Installed)
4 extra/intel-ucode 20230613-1 (5.9 MiB 7.0 MiB) (Installed)
3 extra/intel-metee 3.2.3-1 (15.6 KiB 33.8 KiB) (Installed)
2 extra/intel-gmmlib 22.3.3-1 (257.1 KiB 1.6 MiB) (Installed)
```


> Another issue I've been having is related to the compiler settings here: ![image](https://user-images.githubusercontent.com/7517025/258601103-839b6207-2b1d-479e-92a9-aeb8503e4acb.png) This is an experimental feature and it's not a big deal if it isn't working, but it has regressed on my machine. Perhaps its just some missing feature.

It still works despite those errors but IPEX compile is slower than no compile right now.


>  What does 'IPEX Optimize for Intel GPUs' do? Doesn't seem to make much of a difference on my machine.

It's the `torch.xpu.optimize()` function that Intel recommends to use. 
Sometimes it fixes random NaNs but generally i don't see any difference.
